### PR TITLE
bpo-36876: Make some static string literal arrays constant.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This gives us guarantees about immutability.

https://bugs.python.org/issue36876

### CODE REVIEW

File: Modules/_datetimemodule.c [hunk 1]
```diff
@@ -4012,8 +4012,8 @@ static PyGetSetDef time_getset[] = {
  * Constructors.
  */
 
-static char *time_kws[] = {"hour", "minute", "second", "microsecond",
-                           "tzinfo", "fold", NULL};
+static const char *time_kws[] = {"hour", "minute", "second", "microsecond",
+                                "tzinfo", "fold", NULL};
```
Comment (pganssle) [inline Modules/_datetimemodule.c]: I think this got mis-aligned.

```suggestion
                                 "tzinfo", "fold", NULL};
```

File: Modules/_decimal/_decimal.c [hunk 2]
```diff
@@ -3689,7 +3689,7 @@ dec_##MPDFUNC(PyObject *self, PyObject *dummy UNUSED)       \
 static PyObject *                                                         \
 dec_##MPDFUNC(PyObject *self, PyObject *args, PyObject *kwds)             \
 {                                                                         \
-    static char *kwlist[] = {"context", NULL};                            \
+    static const char *kwlist[] = {"context", NULL};                            \
```
Comment (pganssle) [inline Modules/_decimal/_decimal.c]:
```suggestion
    static const char *kwlist[] = {"context", NULL};                      \
```

**Review by 1st1 — APPROVED** [review]

File: Modules/_decimal/_decimal.c [hunk 3]
```diff
@@ -3706,7 +3706,7 @@ dec_##MPDFUNC(PyObject *self, PyObject *args, PyObject *kwds)             \
 static PyObject *                                              \
 dec_##MPDFUNC(PyObject *self, PyObject *args, PyObject *kwds)  \
 {                                                              \
-    static char *kwlist[] = {"context", NULL};                 \
+    static const char *kwlist[] = {"context", NULL};                 \
```
Comment (pganssle) [inline Modules/_decimal/_decimal.c]:
```suggestion
    static const char *kwlist[] = {"context", NULL};           \
```

File: Modules/_decimal/_decimal.c [hunk 4]
```diff
@@ -3735,7 +3735,7 @@ dec_##MPDFUNC(PyObject *self, PyObject *args, PyObject *kwds)  \
 static PyObject *                                                \
 dec_##MPDFUNC(PyObject *self, PyObject *args, PyObject *kwds)    \
 {                                                                \
-    static char *kwlist[] = {"other", "context", NULL};          \
+    static const char *kwlist[] = {"other", "context", NULL};          \
```
Comment (pganssle) [inline Modules/_decimal/_decimal.c:3738]: Few other alignment issues like the ones above - probably best to align the newline escapes as above.

**Review by pganssle — CHANGES_REQUESTED** [review]
Looks good overall, a few formatting nits around whitespace I found.

File: Modules/_multiprocessing/semaphore.c [hunk 5]
```diff
@@ -435,7 +435,7 @@ semlock_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
     int kind, maxvalue, value, unlink;
     PyObject *result;
     char *name, *name_copy = NULL;
-    static char *kwlist[] = {"kind", "value", "maxvalue", "name", "unlink",
+    static const char *kwlist[] = {"kind", "value", "maxvalue", "name", "unlink",
                              NULL};
```
Comment (pganssle) [inline Modules/_multiprocessing/semaphore.c]:
```suggestion
                                   NULL};
```

I think this is the right number of spaces to re-align this...

File: Modules/mmapmodule.c [hunk 6]
```diff
@@ -1278,7 +1278,7 @@ new_mmap_object(PyTypeObject *type, PyObject *args, PyObject *kwdict)
     HANDLE fh = 0;
     int access = (access_mode)ACCESS_DEFAULT;
     DWORD flProtect, dwDesiredAccess;
-    static char *keywords[] = { "fileno", "length",
+    static const char *keywords[] = { "fileno", "length",
```
Comment (pganssle) [inline Modules/mmapmodule.c:1281]: Next two lines are also misaligned.

File: Modules/posixmodule.c [hunk 7]
```diff
@@ -8976,7 +8976,7 @@ posix_sendfile(PyObject *self, PyObject *args, PyObject *kwdict)
     struct sf_hdtr sf;
     int flags = 0;
     /* Beware that "in" clashes with Python's own "in" operator keyword */
-    static char *keywords[] = {"out", "in",
+    static const char *keywords[] = {"out", "in",
                                 "offset", "count",
```
Comment (pganssle) [inline Modules/posixmodule.c:8980]: Alignment here again.

File: Modules/posixmodule.c [hunk 8]
```diff
@@ -9092,7 +9092,7 @@ posix_sendfile(PyObject *self, PyObject *args, PyObject *kwdict)
 #else
     Py_ssize_t count;
     PyObject *offobj;
-    static char *keywords[] = {"out", "in",
+    static const char *keywords[] = {"out", "in",
```
Comment (pganssle) [inline Modules/posixmodule.c:9095]: Alignment again.

File: Modules/selectmodule.c [hunk 9]
```diff
@@ -1821,7 +1821,7 @@ static int
 kqueue_event_init(kqueue_event_Object *self, PyObject *args, PyObject *kwds)
 {
     PyObject *pfd;
-    static char *kwlist[] = {"ident", "filter", "flags", "fflags",
+    static const char *kwlist[] = {"ident", "filter", "flags", "fflags",
                              "data", "udata", NULL};
```
Comment (pganssle) [inline Modules/selectmodule.c]:
```suggestion
                                   "data", "udata", NULL};
```

File: Python/_warnings.c [hunk 10]
```diff
@@ -1024,7 +1024,7 @@ get_source_line(PyObject *module_globals, int lineno)
 static PyObject *
 warnings_warn_explicit(PyObject *self, PyObject *args, PyObject *kwds)
 {
-    static char *kwd_list[] = {"message", "category", "filename", "lineno",
+    static const char *kwd_list[] = {"message", "category", "filename", "lineno",
                                 "module", "registry", "module_globals",
```
Comment (pganssle) [inline Python/_warnings.c:1028]: Alignment here again.

**Review by pganssle — CHANGES_REQUESTED** [review]
Here are all the alignment issues I found. Hopefully the "suggestions" are the right number of spaces.

**Comment by serhiy-storchaka** [comment]
Changing `char *kwlist[]` to `const char *kwlist[]` is incompatible change, because `PyArg_ParseTupleAndKeywords()` takes `char **` which is not compatible with `const char **`. The compiler should raise warnings or even errors about this. If not this this change would be made years ago.

It is possible to change the API, but it is long a complex process. It is a separate issue.

Adding consts in `Python/Python-ast.c` may be good, but this file is generated. You should modify the generating script.

**Comment by serhiy-storchaka** [comment]
Note also that you can add two `const`s in `static char *something[]`: `static const char * const something[]`. One for C-strings, and other for an array.

See #15824 for changes in `Python/Python-ast.c`.

**Comment by ericsnowcurrently** [comment]
Hmm, this won't work out.

**Comment by georgthegreat** [comment]
At the time there is a bunch of forgotten const / const_cast / `(char**)` c-style casts everywhere in `PyArg_ParseTupleAndKeywords` invocations.

The list includes the following pip packages (but definitely not limited to):

* [brotli](https://github.com/google/brotli/blob/63be8a99401992075c23e99f7c84de1c653e39e2/python/_brotli.cc#L462)
* [rapidjson](https://github.com/python-rapidjson/python-rapidjson/blob/dd69a997b8174be7f933cdd37853a97fe26b1e9e/rapidjson.cpp#L1542)
* [annoy](https://github.com/spotify/annoy/blob/be83fbae5760074e2de084c2d1fdc6e572b0fbaa/src/annoymodule.cc#L147)

Is there any way to continue working on the issue?

**Comment by georgthegreat** [comment]
@serhiy-storchaka, could you, please, elaborate on

> It is possible to change the API, but it is long a complex process. It is a separate issue.)

At the time the lack of const causes problems i. e. when compiling in [MSVC standard conformance](https://docs.microsoft.com/ru-ru/cpp/build/reference/permissive-standards-conformance?view=msvc-160) mode (which is default for `/std:c++latest` / `/std:c++20`).

**Pull request closed without merging**